### PR TITLE
Adding a FNetMaskedLM task model and preprocessor

### PR DESCRIPTION
**Pull request by apupneja**
Solves #722

### CODE REVIEW

**Comment by apupneja** [comment]
Can you please approve the CI for this PR @mattdangerw?

File: keras_nlp/models/f_net/f_net_masked_lm.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    FNetMaskedLMPreprocessor,
+)
+from keras_nlp.models.f_net.f_net_presets import backbone_presets
+from keras_nlp.models.task import Task
+from keras_nlp.utils.python_utils import classproperty
+
+
+@keras.utils.register_keras_serializable(package="keras_nlp")
+class FNetMaskedLM(Task):
+    """An end-to-end FNet model for the masked language modeling task.
```
Comment (mattdangerw) [inline keras_nlp/models/f_net/f_net_masked_lm.py:31]: Looks like the formatting for the docstring may have gotten messed up during a copy paste. Refer to other preprocessing layer docstring as a reference.

Possibly you are copying from the github code browser, which won't properly grab the whitespace.

File: keras_nlp/models/f_net/f_net_masked_lm_preprocessor.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+                f"{self.__class__.__name__} generates `y` and `sample_weight` "
+                "based on your input data, but your data already contains `y` "
+                "or `sample_weight`. Your `y` and `sample_weight` will be "
+                "ignored."
+            )
+        x = super().call(x)
+        token_ids, segment_ids, padding_mask = (
+            x["token_ids"],
+            x["segment_ids"],
+            x["padding_mask"],
```
Comment (mattdangerw) [inline keras_nlp/models/f_net/f_net_masked_lm_preprocessor.py:186]: i'm pretty sure f_net has no padding mask, so I think we will need to remove this

File: keras_nlp/models/f_net/f_net_preprocessor.py [hunk 3]
```diff
@@ -174,6 +174,7 @@ def call(self, x, y=None, sample_weight=None):
         x = {
             "token_ids": token_ids,
             "segment_ids": segment_ids,
+            "padding_mask": token_ids != self.tokenizer.pad_token_id,
```
Comment (mattdangerw) [inline keras_nlp/models/f_net/f_net_preprocessor.py:177]: We don't actually want to do this I believe. f_net, because of the transformations it does to the sequence during it's transformer blocks, does not have a padding mask (though it does have a padding token).

cc @abheesht17 to double check this.

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Thanks for the PR! Left some initial comments.

**Comment by mattdangerw** [comment]
> Can you please approve the CI for this PR @mattdangerw?

Done! And left some initial comments.

File: keras_nlp/models/f_net/f_net_preprocessor.py (hunk 3 above)
Comment (apupneja) [reply keras_nlp/models/f_net/f_net_preprocessor.py:177]: Yes, I believe you're correct. I looked at the official implementation ([here](https://github.com/google-research/google-research/blob/master/f_net/input_pipeline.py): function at line 57). I'll fix that. For the lm_masked version of the preprocessor, I should return the mask positions, right?

**Comment by apupneja** [comment]
I've tested this on the gist you shared in another PR.

Here's the link for it: https://colab.research.google.com/gist/apupneja/8258f796a76940394d4a510ed61d08d7/deberta-masked-lm.ipynb

File: keras_nlp/models/f_net/f_net_preprocessor.py (hunk 3 above)
Comment (abheesht17) [reply keras_nlp/models/f_net/f_net_preprocessor.py]: Sorry, missed this. That is correct. I've confirmed this with the authors of the paper and they stated that they simply mix the entire sequence using the Fourier Transform. The model might be sensitive to examples with very different amounts of padding, but no major issues/problems were observed in practice.

File: keras_nlp/models/f_net/f_net_masked_lm.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    This model can optionally be configured with a `preprocessor` layer, in
+    which case inputs can be raw string features during `fit()`, `predict()`,
+    and `evaluate()`. Inputs will be tokenized and dynamically masked during
+    training and evaluation. This is done by default when creating the model
+    with `from_preset()`.
+
+    Disclaimer: Pre-trained models are provided on an "as is" basis, without
+    warranties or conditions of any kind. The underlying model is provided by a
+    third party and subject to a separate license, available
+    [here](https://github.com/facebookresearch/fairseq).
```
Comment (mattdangerw) [inline keras_nlp/models/f_net/f_net_masked_lm.py:47]: this is not correct for this model! we can use the same disclaimer as BERT

File: keras_nlp/models/f_net/f_net_masked_lm_preprocessor.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    ds = ds.map(preprocessor, num_parallel_calls=tf.data.AUTOTUNE)
+
+    # Alternatively, you can create a preprocessor from your own vocabulary.
+    vocab_data = tf.data.Dataset.from_tensor_slices(
+        ["the quick brown fox", "the earth is round"]
+    )
+
+    # Creating sentencepiece tokenizer for FNet LM preprocessor
+    bytes_io = io.BytesIO()
+
```
Comment (mattdangerw) [inline keras_nlp/models/f_net/f_net_masked_lm_preprocessor.py:99]: we can remove the empty newlines past this point in the code block

File: keras_nlp/models/f_net/f_net_masked_lm_preprocessor.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        bos_piece="[CLS]",
+        eos_piece="[SEP]",
+        user_defined_symbols="[MASK]",
+    )
+
+    proto = bytes_io.getvalue()
+
+    tokenizer = FNetTokenizer(proto=proto)
+
+    preprocessor = FNetMaskedLMPreprocessor(
```
Comment (mattdangerw) [inline keras_nlp/models/f_net/f_net_masked_lm_preprocessor.py:120]: format this as one line

**Review by mattdangerw — APPROVED** [review]
Thank you! This looks great! Just a few small comments to clean up, and this needs a merge/rebase with the master branch.

**Comment by mattdangerw** [comment]
Also thanks so much for testing this our via the gist! Super helpful.

File: keras_nlp/models/f_net/f_net_masked_lm_preprocessor.py (hunk 6 above)
Comment (apupneja) [reply keras_nlp/models/f_net/f_net_masked_lm_preprocessor.py:120]: By one line, do you mean that I should create the object inside the `FNetMaskedLMPreprocessor` constructor parameters?

`preprocessor = FNetMaskedLMPreprocessor(
        tokenizer=FNetTokenizer(proto=proto)
 )`

Something like this?

Comment (mattdangerw) [reply keras_nlp/models/f_net/f_net_masked_lm_preprocessor.py:115]: Oh I just mean what inside the parentheses here.
`preprocessor = FNetMaskedLMPreprocessor(tokenizer=tokenizer)`

Comment (apupneja) [reply keras_nlp/models/f_net/f_net_masked_lm_preprocessor.py]: Right. Fixed it.

**Comment by mattdangerw** [comment]
Thanks!
